src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -145,164 +160,115 @@ public void Shutdown()
+                        }
                     }
-                    else
+                    catch (OperationCanceledException ex) when (ex.CancellationToken == queueItem.cancellationToken)
                     {
-                        // Non mutating are fire-and-forget because they are by definition readonly. Any errors
-                        // will be sent back to the client but we can still capture errors in queue processing
-                        // via NFW, though these errors don't put us into a bad state as far as the rest of the queue goes.
-                        // Furthermore we use Task.Run here to protect ourselves against synchronous execution of work
-                        // blocking the request queue for longer periods of time (it enforces parallelizabilty).
-                        _ = Task.Run(() => ExecuteCallbackAsync(work, context, _cancelSource.Token), _cancelSource.Token).ReportNonFatalErrorAsync();
+                        // Explicitly ignore this exception as cancellation occured as a result of our linked cancellation token.
+                        // This means either the queue is shutting down or the request itself was cancelled.
+                        //   1.  If the queue is shutting down, then while loop will exit before the next iteration since it checks for cancellation.
+                        //   2.  Request cancellations are normal so no need to report anything there.
                     }
                 }
             }
-            catch (OperationCanceledException e) when (e.CancellationToken == _cancelSource.Token)
-            {
-                // If cancellation occurs as a result of our token, then it was either because we cancelled it in the Shutdown
-                // method, if it happened during a mutating request, or because the queue was completed in the Shutdown method
-                // if it happened while waiting to dequeue the next item. Either way, we're already shutting down so we don't
-                // want to log it.
-            }
-            catch (Exception e) when (FatalError.ReportAndCatch(e))
+            catch (Exception ex) when (FatalError.ReportAndCatch(ex))
             {
-                _logger.TraceException(e);
-
-                // If there was an in progress work item that failed in the queue logic, set the result of the queue item
-                // to the exception so that it bubbles back to the caller.
-                inProgressWorkItem?.HandleQueueFailure(e);
-                OnRequestServerShutdown($"Error occurred processing queue in {_serverName}: {e.Message}.");
+                // We encountered an unexpected exception in processing the queue or in a mutating request.
+                // Log it, shutdown the queue, and exit the loop.
+                _logger.TraceException(ex);
+                OnRequestServerShutdown($"Error occurred processing queue in {_serverName}: {ex.Message}.");
+                return;
```

Comment:
nice.  so it worked to not have to tell the item to handle this?

Reply:
yup, discussed https://github.com/dotnet/roslyn/pull/57617#discussion_r745263455 offline